cpp/lib/Authenticator.cpp

```diff
@@ -0,0 +1,359 @@
+    }
+
+    AuthenticatorType auth_type = getAuthenticatorType(conn->authenticator);
+    try
+    {
+      if (AUTH_JWT == auth_type)
+      {
+        conn->auth_object = static_cast<Snowflake::Client::IAuthenticator*>(
+                              new Snowflake::Client::AuthenticatorJWT(conn));
+      }
+    }
+    catch (...)
+    {
+      SET_SNOWFLAKE_ERROR(&conn->error, SF_STATUS_ERROR_GENERAL,
+                          "authenticator initialization failed",
+                          SF_SQLSTATE_GENERAL_ERROR);
+      return SF_STATUS_ERROR_GENERAL;
+    }
+
+    return SF_STATUS_SUCCESS;
+  }
+
+  int64 auth_get_renew_timeout(SF_CONNECT * conn)
+  {
+    if (!conn || !conn->auth_object)
+    {
+      return 0;
+    }
+
+    try
+    {
+      return static_cast<Snowflake::Client::IAuthenticator*>(conn->auth_object)
+               ->getAuthRenewTimeout();
+    }
+    catch (...)
+    {
+      return 0;
+    }
+  }
+
+  SF_STATUS STDCALL auth_authenticate(SF_CONNECT * conn)
+  {
+    if (!conn || !conn->auth_object)
+    {
+      return SF_STATUS_SUCCESS;
+    }
+
+    try
+    {
+      static_cast<Snowflake::Client::IAuthenticator*>(conn->auth_object)->authenticate();
+    }
+    catch (...)
+    {
+      SET_SNOWFLAKE_ERROR(&conn->error, SF_STATUS_ERROR_GENERAL,
+                          "authentication failed",
+                          SF_SQLSTATE_GENERAL_ERROR);
+      return SF_STATUS_ERROR_GENERAL;
+    }
+
+    return SF_STATUS_SUCCESS;
+  }
+
+  void auth_update_json(SF_CONNECT * conn, cJSON* body)
+  {
+    if (!conn || !conn->auth_object)
+    {
+      return;
+    }
+
+    try
+    {
+      static_cast<Snowflake::Client::IAuthenticator*>(conn->auth_object)->
+        updateDataMap(body);
+    }
+    catch (...)
+    {
+      ; // Do nothing
+    }
+
+    return;
+  }
+
+  void auth_renew_json(SF_CONNECT * conn, cJSON* body)
+  {
+    if (!conn || !conn->auth_object)
+    {
+      return;
+    }
+
+    try
+    {
+      static_cast<Snowflake::Client::IAuthenticator*>(conn->auth_object)->
+        renewDataMap(body);
+    }
+    catch (...)
+    {
+      ; // Do nothing
+    }
+
+    return;
+  }
+
+  void STDCALL auth_terminate(SF_CONNECT * conn)
+  {
+    if (!conn || !conn->auth_object)
+    {
+      return;
+    }
+
+    AuthenticatorType auth_type = getAuthenticatorType(conn->authenticator);
+    try
+    {
+      if (AUTH_JWT == auth_type)
+      {
+        delete static_cast<Snowflake::Client::AuthenticatorJWT*>(conn->auth_object);
+      }
+    }
+    catch (...)
+    {
+      ; // Do nothing
+    }
+
+    return;
+  }
+
+} // extern "C"
+
+namespace Snowflake
+{
+namespace Client
+{
+
+  IAuthenticator *IAuthenticator::buildAuthenticator(AuthenticatorType type, SF_CONNECT *conn)
+  {
+    switch(type) {
+      case AUTH_JWT:
+        return new AuthenticatorJWT(conn);
+      default:
+        return nullptr;
+    }
+  }
+
+  void IAuthenticator::renewDataMap(cJSON *dataMap)
```

Comment:
@SimbaGithub can you please remind me under what scenario we'd need this?

Reply:
Nvm .. I remember this is for renewing Token correct?

Reply:
Yes, it's for renewing jwt token.